www/js/lib/popovers.js

```diff
@@ -97,6 +133,95 @@ function getArticleLede (href, baseUrl, articleDocument, archive) {
+                redirectPath = namespace + redirectPath;
+            }
+            
+            return archive.callLibzimWorker({ 
+                action: 'getEntryByPath', 
+                path: redirectPath,
+                follow: false
+            }).then(function (finalRet) {
+                if (!finalRet || !finalRet.content) {
+                    throw new Error('No content found for redirect: ' + zimURL);
+                }
+                const htmlArticle = finalRet.content instanceof Uint8Array ? 
+                    archive.getUtf8FromData(finalRet.content) : finalRet.content;
+                
+                        return parseArticleContentForBalloon(htmlArticle, redirectPath, articleDocument);
+            });
+        } else {
+            // No redirect, process the content normally
+            if (!ret || !ret.content) {
+                throw new Error('No content found for ' + zimURL);
+            }
+            const htmlArticle = ret.content instanceof Uint8Array ? 
+                archive.getUtf8FromData(ret.content) : ret.content;
+            
+            // Use the original zimURL as the path
+            return parseArticleContentForBalloon(htmlArticle, zimURL, articleDocument);
+        }
+    });
+}
+
+// Helper function to parse article content and create balloon
+function parseArticleContentForBalloon(htmlArticle, zimURL, articleDocument) {
+    const parser = new DOMParser();
+    const doc = parser.parseFromString(htmlArticle, 'text/html');
+    const articleBody = doc.body;
+    
+    if (articleBody) {
+        // Establish the popup balloon's base URL and the absolute path for calculating the ZIM URL of links and images
+        const balloonBaseURL = zimURL.replace(/[^/]+$/, '');
+        const docUrl = new URL(articleDocument.location.href);
+        const rootRelativePathPrefix = docUrl.pathname.replace(/(.*\.zim[^/]*\/).*$/i, '$1');
+        
+        // Clean up the lede content
+        const nonEmptyParagraphs = cleanUpLedeContent(articleBody);
+        // Concatenate paragraphs to fill the balloon
+        let balloonString = '';
+        if (nonEmptyParagraphs.length > 0) {
+            balloonString = fillBalloonString(nonEmptyParagraphs, balloonBaseURL, rootRelativePathPrefix);
+        }
+        // If we have a lede, we can now add an image to the balloon, but only if we are in ServiceWorker mode
+        if (balloonString && params.contentInjectionMode === 'serviceworker') {
+            const imageHTML = getImageHTMLFromNode(articleBody, balloonBaseURL, rootRelativePathPrefix);
+            if (imageHTML) {
+                balloonString = imageHTML + balloonString;
+            }
+        }
+        if (!balloonString) {
+            throw new Error('No article lede or image');
+        } else {
+            return balloonString;
+        }
+    } else {
+        throw new Error('No article body found');
+    }
+}
```

Comment:
As I understand it, this entire function was moved from its previous position, as we had discussed. However, you have changed some things, and in particular the regex for matching rootRelativePathPrefix. Please don't alter existing code unless you really need to. Please go through this and either restore altered code or explain why you needed to change it for your implementation, as it is a shared helper function.